imap_processing/ultra/l0/decom_tools.py

```diff
@@ -0,0 +1,255 @@
+    Software Specification document.
+    """
+    rows = 54
+    cols = 180
+    pixels_per_block = 15
+
+    blocks_per_row = cols // pixels_per_block
+
+    # Compressed pixel matrix
+    p = np.zeros((rows, cols), dtype=np.uint16)
+    # Decompressed pixel matrix
+    p_decom = np.zeros((rows, cols), dtype=np.uint16)
+
+    pos = 0  # Starting position in the binary string
+
+    for i in range(rows):
+        for j in range(blocks_per_row):
+            # Read the width for the block.
+            w, pos = read_and_advance(binary_data, width_bit, pos)
+            for k in range(pixels_per_block):
+                # Handle the special case in which the width is 0
+                if w == 0:
+                    value = 0
+                else:
+                    # Find the value of each pixel in the block
+                    value, pos = read_and_advance(binary_data, w, pos)
+
+                # if the least significant bit of value is set (odd)
+                if value & 0x01:
+                    # value >> 1: shifts bits of value one place to the right
+                    # ~: bitwise NOT operator (flips bits)
+                    delta_f = ~(value >> 1)
+                else:
+                    delta_f = value >> 1
+
+                # Calculate the new pixel value and update pixel0
+                column_index = j * pixels_per_block + k
+                # 0xff is the hexadecimal representation of the number 255,
+                # Keeps only the last 8 bits of the result of pixel0 - delta_f
+                # This operation ensures that the result is within the range
+                # of an 8-bit byte (0-255)
+                p[i][column_index] = (pixel0 - delta_f) & 0xFF
+                # Perform logarithmic decompression on the pixel value
+                p_decom[i][column_index] = log_decompression(
+                    p[i][column_index], mantissa_bit_length
+                )
+                pixel0 = p[i][column_index]
+        pixel0 = p[i][0]
+
+    return p_decom
+
+
+def read_image_raw_events_binary(packet, decom_data: dict):
+    """Convert contents of binary string 'EVENTDATA' into values.
+
+    Parameters
+    ----------
+    packet : space_packet_parser.parser.Packet
+        Packet.
+    decom_data : dict
+        Parsed data.
+
+    Returns
+    -------
+    decom_data : dict
+        Each for loop appends to the existing dictionary.
+    """
+    binary = packet.data["EVENTDATA"].raw_value
+    count = packet.data["COUNT"].derived_value
+    event_length = int(len(binary) / count) if count else 0
+
+    # Uses fill value for all packets that do not contain event data.
+    if count == 0:
+        append_fillval(decom_data, packet)
+
+    # For all packets with event data, parses the binary string
+    else:
+        for i in range(count):
+            start_index = i * event_length
+            event_binary = binary[start_index : start_index + event_length]
```

Comment:
Are all events same length? just curious

Reply:
Yes. Also 166 bits or 0 (if there is no data).

<xtce:Parameter name="EVENTDATA" parameterTypeRef="eventdata">
	<xtce:LongDescription>Event Data, 166 bits per event</xtce:LongDescription>
</xtce:Parameter>